Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP1Connections.swift

```diff
@@ -0,0 +1,602 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIOCore
+import XCTest
+
+extension HTTPConnectionPool {
+    /// Represents the state of a single HTTP/1.1 connection
+    private struct HTTP1ConnectionState {
+        enum State {
+            /// the connection is creating a connection. Valid transitions are to: .backingOff, .available and .failed
+            case starting
+            /// the connection is waiting to retry the establishing a connection. Transition to .closed. Must be replaced for a retry.
+            case backingOff
+            /// the connection is available for a new request. Valid transitions to: .leased and .closed
+            case available(Connection, since: NIODeadline)
```

Comment:
"idle" and "available" are both used through this PR to refer to the same thing, can we stick to one?